src/main/scala/sparkz/core/network/peer/PeerManager.scala

```diff
@@ -137,42 +135,63 @@ object PeerManager {
     case class SeenPeers(howMany: Int) extends GetPeers[Seq[PeerInfo]] {
 
       override def choose(knownPeers: Map[InetSocketAddress, PeerInfo],
+                          peers: Map[InetSocketAddress, PeerInfo],
                           blacklistedPeers: Seq[InetAddress],
                           sparkzContext: SparkzContext): Seq[PeerInfo] = {
-        val recentlySeenNonBlacklisted = knownPeers.values.toSeq
+        val recentlySeenKnownPeers = knownPeers.values.toSeq
+          .filter { p =>
+            p.connectionType.isDefined || p.lastHandshake > 0
+          }
+        val recentlySeenNonBlacklisted = peers.values.toSeq
           .filter { p =>
             (p.connectionType.isDefined || p.lastHandshake > 0) &&
               !blacklistedPeers.exists(ip => p.peerSpec.declaredAddress.exists(_.getAddress == ip))
           }
-        Random.shuffle(recentlySeenNonBlacklisted).take(howMany)
+        Random.shuffle(recentlySeenKnownPeers ++ recentlySeenNonBlacklisted).take(howMany)
       }
     }
 
     case object GetAllPeers extends GetPeers[Map[InetSocketAddress, PeerInfo]] {
 
       override def choose(knownPeers: Map[InetSocketAddress, PeerInfo],
+                          peers: Map[InetSocketAddress, PeerInfo],
                           blacklistedPeers: Seq[InetAddress],
-                          sparkzContext: SparkzContext): Map[InetSocketAddress, PeerInfo] = knownPeers
+                          sparkzContext: SparkzContext): Map[InetSocketAddress, PeerInfo] = knownPeers ++ peers
     }
 
-    case class RandomPeerExcluding(excludedPeers: Seq[Option[InetSocketAddress]]) extends GetPeers[Option[PeerInfo]] {
+    case class RandomPeerForConnectionExcluding(excludedPeers: Seq[Option[InetSocketAddress]]) extends GetPeers[Option[PeerInfo]] {
       private val secureRandom = new SecureRandom()
 
       override def choose(knownPeers: Map[InetSocketAddress, PeerInfo],
```

Comment:
This implementation pushes knownPeers first, since they should be privileged peers, and ignores other peers if any known peer is found. This opens up a corner-case scenario where we have at least one known peer we can't connect to because it's offline, and this method keeps returning that peer, since it's a knownPeer and we're not connected to it.